### DISCUSSION TITLE
Variable with validation causes terraform_unused_declarations in 0.49.0

### DISCUSSION THREAD

**Discussion by gdsotirov** [terraform-linters/tflint#1910] [General]
I have a Terraform module defining common variables and their validation, which is used by all other modules (through a symbolic link). For example, the module looks like the following:

```tf
terraform {
	# Output values precondition requires Terraform 1.2 or later
	required_version	= ">= 1.2"
}

locals {
	# See https://cloud.google.com/compute/docs/regions-zones#available
	gcp_regions				= [
		"asia-east1",				# Changhua County, Taiwan
		"asia-east2",				# Hong Kong
		# and so on...
	]
}

variable "region" {
	default		= null
	description	= "Region where resources are located"
	type		= string

	validation {
		condition		= ( var.region == null || can(regex("^[a-z]+-[a-z]+[0-9]$", var.region)) )
		error_message	= format("Region does not seem valid!")
	}
}

output "region_validation" {
	value	= null

	precondition {
		condition		= var.region == null || contains(local.gcp_regions, coalesce(var.region, "n/a"))
		error_message	= format("Invalid region '%s'!", coalesce(var.region, "n/a"))
	}
}

variable "zone" {
	default		= null
	description = "Zone to use for zonal resources."
	type		= string

	validation {
		# See https://github.com/hashicorp/terraform/issues/25609
		condition		= ( var.zone == null || can(regex("^[a-z]+-[a-z]+][0-9]-[a-f]$", var.zone)) )
		error_message	= format("Zone does not seem valid!")
	}
}
```

In other words all variables have [validation](https://developer.hashicorp.com/terraform/language/values/variables#custom-validation-rules) and some variables have more complex validation through [output precondition](https://developer.hashicorp.com/terraform/language/expressions/custom-conditions#preconditions-and-postconditions). With tflint 0.48.0 this code was not producing any error or warning, but with 0.49.0 I now have:

```
Warning: [Fixable] variable "zone" is declared but not used (terraform_unused_declarations)

  on inputs.tf line 28:
  28: variable "zone" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.5.0/docs/rules/terraform_unused_declarations.md
```

Although, I find nothing about this change in [the release notes](https://github.com/terraform-linters/tflint/releases/tag/v0.49.0), I see the reason for it, e.g. using variable in output is _real_ usage, but I would like to have the previous behavior without having to disable the warning. What do you think?

**Comment by wata727** [comment]
This is due to an update the bundled plugin to v0.5.0. https://github.com/terraform-linters/tflint/pull/1883
See https://github.com/terraform-linters/tflint-ruleset-terraform/pull/133 for more information about this change.

> but I would like to have the previous behavior without having to disable the warning. What do you think?

Unfortunately, you cannot reproduce the same behavior as before without disabling the warning, but if you want to leave the variable for future use, the `tflint-ignore` annotation may be useful.

**Reply by gdsotirov** [reply]
OK. Clear. It's well visible in [0.5.0's release notes](https://github.com/terraform-linters/tflint-ruleset-terraform/releases/tag/v0.5.0). I supposed something like this and I do understand the reason. However, for my case I really do not want to repeat a number common variable declarations (e.g. `region`, `zone`, etc. as in the example above) in even larger number of modules that need them, so I worked it around by adding null output with precondition check (as `region_validation` from the example) for each common variable. :-)
